blscurve/blst/bls_sig_io.nim

```diff
@@ -41,44 +41,111 @@ func toHex*(
 
 func fromBytes*(
        obj: var (Signature|ProofOfPossession),
-       raw: openarray[byte] or array[96, byte]
+       raw: array[96, byte] or array[192, byte]
       ): bool {.inline.} =
   ## Initialize a BLS signature scheme object from
   ## its raw bytes representation.
   ## Returns true on success and false otherwise
-  const L = 96
-  when raw is array:
-    result = obj.point.blst_p2_uncompress(raw) == BLST_SUCCESS
-  else:
-    if raw.len != L:
-      return false
-    let pa = cast[ptr array[L, byte]](raw[0].unsafeAddr)
-    result = obj.point.blst_p2_uncompress(pa[]) == BLST_SUCCESS
+  result =
+    when raw.len == 96:
+      obj.point.blst_p2_uncompress(raw) == BLST_SUCCESS
+    elif raw.len == 192:
+      obj.point.blst_p2_deserialize(raw) == BLST_SUCCESS
+    else: false
+
   # Infinity signatures are allowed if we receive an empty aggregated signature
   if result:
     result = bool obj.point.blst_p2_affine_in_g2()
 
+func fromBytesKnownOnCurve*(
```

Comment:
what happens if the point is not on the curve?

Reply:
Bad things, basically we don't work on the full curve but only in a subgroup of the curve.
To ensure every pubkey or signature are in that subgroup we clear the cofactor.
If we miss clearing the cofactor we open ourselves to a lot of mathematical vulnerabilities.
There was a large incident with Monero in 2017.

- https://www.getmonero.org/2017/05/17/disclosure-of-a-major-bug-in-cryptonote-based-currencies.html
- https://jonasnick.github.io/blog/2017/05/23/exploiting-low-order-generators-in-one-time-ring-signatures/
- https://loup-vaillant.fr/tutorials/cofactor

So this is to be used only in specific cases, I think it was @tersec or @cheatfate who had a case where something was serialized (maybe the db) known on curve and we could avoid the price of the check.